### PR TITLE
Sleep: Allow fractional seconds, fix Kernel timing, fix overflow in Kernel

### PR DESCRIPTION
**Pull request by BenWiederhake**
My masterplan behind strtod: Supporting fractional seconds in `sleep`! :^)

This changes three interrelated things:
- Kernel: No longer overflow on large sleeps, which used to cause hard-to-predict actual delays
- Kernel: Fix "ticks_per_seconds" handling, because some code used to assume that there are always exactly 1000 ticks in a second.
- `/bin/sleep`: Support fractional seconds (even `inf`!)

Here's what it looked like intermediately:

Bildschirmfoto_2020-05-11_15-09-16

(I removed the debug output before pushing.)

### CODE REVIEW

File: Userland/sleep.cpp [hunk 1]
```diff
@@ -24,16 +24,48 @@
...
+// The maximum integer we can definitely exactly represent as a double (2^52).
+// Note that clock_nanosleep can handle even larger numbers (up to around 2^63),
+// but I think it's fine to cap silently at a hundred million years.
+static const double MAX_SLEEPABLE = 4503599627370496;
+
 void handle_sigint(int)
 {
 }
 
+double parse_amount(char* arg, bool& ok) {
```
Comment (awesomekling) [inline Userland/sleep.cpp:44]: Please use `clang-format`.

File: Userland/sleep.cpp [hunk 2]
```diff
@@ -24,16 +24,48 @@
...
+    // This can't go into AK/* because AK can't run anything from stdlib,
+    // so strtod is not available yet.
+    // TODO: Find a way to put this function into AK/
```
Comment (awesomekling) [inline Userland/sleep.cpp:47]: How about we move the bulk of `strtod()` into `AK` and make LibC's `strtod()` a thin wrapper around it? :)

File: Userland/sleep.cpp [hunk 3]
```diff
@@ -42,22 +74,46 @@ int main(int argc, char** argv)
     }
 
     if (argc != 2) {
-        printf("usage: sleep <seconds>\n");
+        printf("Usage:\n        sleep <seconds>\n");
+        printf("\nArguments:\n        seconds Fractional amount of seconds, e.g. 3, 0.01, 1337.42, or inf. The number will be rounded down to the nearest possible amount (usually by 1 ms).\n");
```
Comment (awesomekling) [inline Userland/sleep.cpp:78]: \<sergey\>Please use `Core::ArgsParser`:smile:\</sergey\>

File: Userland/sleep.cpp (hunk 1 above)
Comment (BenWiederhake) [reply Userland/sleep.cpp:44]: Right! I forgot it here.

File: Userland/sleep.cpp (hunk 2 above)
Comment (BenWiederhake) [reply Userland/sleep.cpp:47]: Sure! I thought that would go beyond the scope of a single PR.

File: Userland/sleep.cpp (hunk 3 above)
Comment (BenWiederhake) [reply Userland/sleep.cpp:78]: I wasn't sure whether it's overkill – looking at that string literal again, now I think `ArgsParser` is a better idea indeed!

File: Userland/sleep.cpp (hunk 2 above)
Comment (awesomekling) [reply Userland/sleep.cpp:47]: Yeah you can do that in a separate PR ofc :)

Comment (elcuco) [reply Userland/sleep.cpp:47]: I would like to continue this issue. Code porting to master is quite trivial. Problem is the arg parsing, which is being worked on in https://github.com/SerenityOS/serenity/pull/2867. 

Technically - how to proceed ? Should I create a new PR for this issue, or my PR should be against this PR, and @BenWiederhake  should handle it? (I think closing this one and me handling it giving all credit to @BenWiederhake  is easier)

**Comment by BenWiederhake** [comment]
I removed the commits that touch `sleep` from this commit. For history purposes, the worktree on my harddrive is now available as the tag [historic/dev-sleep-fractional-PR-2192](https://github.com/BenWiederhake/serenity/tree/historic/dev-sleep-fractional-PR-2192).

The remaining commits fix stuff in the Kernel, and seem to have been accepted without issue.

So I suggest that this PR (now without the commits that touch `sleep`, and freshly rebased for conflict-freeness) gets merged, and then @elcuco can proceed however desired. :)

**Comment by elcuco** [comment]
https://github.com/SerenityOS/serenity/pull/2867 - as soon as this gets merged, I will make another PR for restoring support for `double` sleeping in the userspace utility (should be trivial, nah?)

File: Userland/sleep.cpp (hunk 3 above)
Comment (elcuco) [reply Userland/sleep.cpp]: Has been done in 62283ade912bbe0c2ff92509a0ed2a787f330853 - no longer relevant.

**Pull request merged**